toc/rfc/rfc-draft-bindable-secrets.md

```diff
@@ -0,0 +1,293 @@
+
+### What About User-Provided Services?
+
+Why can't developers use user-provided service instances (UPSIs) and service bindings for file-based app configuration? Some apps may have file-based config that don't naturally fit into the service binding model. For example, if an app relies on a config file at a certain path, that plath may not be configurable to `$SERVICE_BINDING_ROOT`. This is especially the case for commercial off-the-shelf software (COTS).
+
+## Proposal
+
+Add two new resources to the V3 Cloud Foundry API: `secret` and the `secret_binding`, along with associated CRUD API endpoints.
+
+### Why Top-Level Resources?
+
+Why not represent secrets as sub-resources of apps like environment variables? Secrets may be shared across multiple apps. For example, a "logical app" may actually be several Cloud Foundry apps resources that need to share certain configuration, certificates, or other credentials. By scoping the `secret` at the `space` level, we can easily support this. Similarly, an app using the blue-green deployment pattern can easily share configuration between versions.
+
+### `secret` and `secret_binding` Resources
+The `secret` will be a `space`-scoped resource (similar to a `route` or a `service_instance`) that can be bound to one or more `app`s in the `space` using a `secret_binding`.
+
+#### Example `secret` Object
+
+```json
+{
+  "guid": "4f00b75b-3455-48f1-ba4a-f3b82ebad943",
+  "created_at": "2020-03-10T15:49:29Z",
+  "updated_at": "2020-03-10T15:49:29Z",
+  "name": "my-secret",
+  "type": "opaque",
+  "credhub_credential_id": null,
+  "relationships": {
+    "space": {
+      "data": {
+        "guid": "5a84d315-9513-4d74-95e5-f6a5501eeef7"
+      }
+    }
+  },
+  "metadata": {
+    "labels": {},
+    "annotations": {}
+  },
+  "links": {
+    "self": {
+      "href": "https://api.example.org/v3/service_instances/c89b3280-fe8d-4aa0-a42e-44465bb1c61c"
+    },
+    "space": {
+      "href": "https://api.example.org/v3/spaces/5a84d315-9513-4d74-95e5-f6a5501eeef7"
+    },
+    "secret_bindings": {
+      "href": "https://api.example.org/v3/secret_bindings?secret_guids=4f00b75b-3455-48f1-ba4a-f3b82ebad943"
+    },
+    "value": {
+      "href": "https://api.example.org/v3/secrets/4f00b75b-3455-48f1-ba4a-f3b82ebad943/value"
+    }
+  }
+}
+```
+
+A `secret` has a type that identifies its contents. To start we will support the `opaque` credential type (corresponds to the `json` [CredHub credential type](https://docs.cloudfoundry.org/credhub/credential-types.html)) which is designed to hold arbitrary user-defined data. In the future, this can be expanded to support more semantic secret types such as those [supported by Kubernetes](https://kubernetes.io/docs/concepts/configuration/secret/#secret-types) or other CredHub-supported types like `certificate`. `opaque` secrets consist of a JSON object whose keys map to files and values map to file content.
```

Comment:
**question:** Can keys only be filenames or also relative paths?

For example:
- The secret binding's `mount_path` is `/etc/my-conf/`.
- The secret's `value` is `{"passwords/db_user.txt":"abc123","certificates/db_cert.pem":"..."}`.

Would result in:
```
etc
└── my-conf
    ├── certificates
    │   └── db_cert.pem
    └── password
        └── db_user.txt
```

Reply:
I feel like supporting relative paths adds complexity that could be resolved by just using multiple secrets. For example, I'd expect many config files might contain JSON and it would be nice if we could avoid parsing the values themselves and just treat it as opaque file contents.